### PR TITLE
IGNITE-7918: memory leak was fixed. Unit test was added.

### PR DESCRIPTION
**Pull request by aealeksandrov**
Some caches didn't take into account that topology minor version could be changed during creating/destroying of the caches dynamically. These places were fixed.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java [hunk 1]
```diff
@@ -724,27 +722,13 @@ private void awaitTopologyVersion(AffinityTopologyVersion topVer) {
...
-
-        if (rmvCnt <= 0) {
-            if (fullSize > MAX_HIST_SIZE * 2)
-                rmvCnt = MAX_HIST_SIZE;
-        }
-
-        if (rmvCnt > 0) {
+    private void onHistoryAdded() {
+        if (fullHistSize.incrementAndGet() > MAX_HIST_SIZE) {
             Iterator<HistoryAffinityAssignment> it = affCache.values().iterator();
```
Comment (ilantukh) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java]: According to our code style, there should be blank lines between code lines within the same { } block.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessor.java [hunk 2]
```diff
@@ -212,6 +218,34 @@ public int partition0(String cacheName, Object key, @Nullable AffinityInfo aff)
...
+        int oldSize = affMap.size();
+
+        for (Map.Entry<AffinityAssignmentKey, IgniteInternalFuture<AffinityInfo>> entry : affMap.entrySet()) {
+            assert entry.getValue() != null;
+
+            if (!entry.getValue().isDone())
+                continue;
+
+            if (topVerRmv.contains(entry.getKey().topVer) || !caches.contains(entry.getKey().cacheName))
+                rmv.add(entry.getKey());
```
Comment (ilantukh) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessor.java]: Why don't you simply use iterator.remove()?

File: modules/core/src/test/java/org/apache/ignite/cache/affinity/AffinityHistoryCleanupTest.java [hunk 3]
```diff
@@ -89,34 +89,34 @@ public void testAffinityHistoryCleanup() throws Exception {
 
             Ignite ignite = startGrid(0);
 
-            checkHistory(ignite, F.asList(topVer(1, 0)), 1);
+            checkHistory(ignite, F.asList(topVer(1, 0)), 1); //fs = 1
```
Comment (ilantukh) [inline modules/core/src/test/java/org/apache/ignite/cache/affinity/AffinityHistoryCleanupTest.java]: What does "fs" mean? Please use more detailed comments.

Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/cache/affinity/AffinityHistoryCleanupTest.java]: I mean FullHistorySize. I will rename it to something more clear.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java (hunk 1 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityAssignmentCache.java]: done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessor.java (hunk 2 above)
Comment (aealeksandrov) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessor.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                    ", expLimit=" + MAX_HIST_SIZE * 3 + "]", size < MAX_HIST_SIZE * 3);
+            }
+            catch (Exception e) {
+                fail("Error was handled [" + e.getMessage() + "]");
+            }
+        }
+        while (count-- > 0);
+    }
+
+    private static IgniteCache<String, String> createLocalCache(Ignite ignite, long id) {
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: javadoc is missing

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        final String cacheName = "localCache" + id;
+        final CacheConfiguration<String, String> cCfg = new CacheConfiguration<>();
+        cCfg.setName(cacheName);
+        cCfg.setCacheMode(CacheMode.LOCAL);
+        cCfg.setGroupName("some group");
+        ignite.destroyCache(cacheName); // local cache is not really local - reference can be kept by other nodes if restart during the load happens
+        return ignite.createCache(cCfg).withKeepBinary();
+    }
+
+    private static IgniteCache<String, String> getOrCreateGlobalCache(Ignite ignite) {
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: javadoc is missing

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        cCfg.setCacheMode(CacheMode.PARTITIONED);
+        cCfg.setOnheapCacheEnabled(false);
+        cCfg.setCopyOnRead(false);
+        cCfg.setBackups(0);
+        cCfg.setWriteBehindEnabled(false);
+        cCfg.setReadThrough(false);
+        return ignite.getOrCreateCache(cCfg).withKeepBinary();
+    }
+
+    private static IgniteDataStreamer<String, String> createGlobalStreamer(Ignite ignite,
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: javadoc is missing

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        IgniteCache<String, String> cache) {
+        IgniteDataStreamer<String, String> streamer = ignite.dataStreamer(cache.getName());
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: newline after declaration

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        streamer.allowOverwrite(true);
+        streamer.skipStore(true);
+        streamer.keepBinary(false);
+        return streamer;
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: newline before return

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    private static IgniteCache<String, String> createLocalCache(Ignite ignite, long id) {
+        final String cacheName = "localCache" + id;
+        final CacheConfiguration<String, String> cCfg = new CacheConfiguration<>();
+        cCfg.setName(cacheName);
+        cCfg.setCacheMode(CacheMode.LOCAL);
+        cCfg.setGroupName("some group");
+        ignite.destroyCache(cacheName); // local cache is not really local - reference can be kept by other nodes if restart during the load happens
+        return ignite.createCache(cCfg).withKeepBinary();
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: newline before return

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        final CacheConfiguration<String, String> cCfg = new CacheConfiguration<>();
+        cCfg.setName(cacheName);
+        cCfg.setStoreKeepBinary(true);
+        cCfg.setCacheMode(CacheMode.PARTITIONED);
+        cCfg.setOnheapCacheEnabled(false);
+        cCfg.setCopyOnRead(false);
+        cCfg.setBackups(0);
+        cCfg.setWriteBehindEnabled(false);
+        cCfg.setReadThrough(false);
+        return ignite.getOrCreateCache(cCfg).withKeepBinary();
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: newline before return

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        cCfg.setName(cacheName);
+        cCfg.setCacheMode(CacheMode.LOCAL);
+        cCfg.setGroupName("some group");
+        ignite.destroyCache(cacheName); // local cache is not really local - reference can be kept by other nodes if restart during the load happens
+        return ignite.createCache(cCfg).withKeepBinary();
+    }
+
+    private static IgniteCache<String, String> getOrCreateGlobalCache(Ignite ignite) {
+        final String cacheName = "GlobalCache";
+        final CacheConfiguration<String, String> cCfg = new CacheConfiguration<>();
```
Comment (dspavlov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: newline after decl

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 4 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 5 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 6 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 7 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 8 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 9 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 10 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java (hunk 11 above)
Comment (aealeksandrov) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessorMemoryLeakTest.java]: fixed

**Pull request closed without merging**